p4_symbolic/ir/ir.cc

```diff
@@ -0,0 +1,295 @@
+    RETURN_IF_ERROR(ValidateHeaderTypeFields(unparsed_field));
+
+    HeaderField &field =
+        (*output->mutable_fields())[unparsed_field.values(0).string_value()];
+    field.set_name(unparsed_field.values(0).string_value());
+    field.set_bitwidth(unparsed_field.values(1).number_value());
+    field.set_signed_(unparsed_field.values(2).bool_value());
+    field.set_header_type(header.name());
+  }
+
+  return absl::OkStatus();
+}
+
+// Parsing values.
+absl::Status TransformLValue(const google::protobuf::Value &bmv2_value,
+                             const std::vector<std::string> &variables,
+                             LValue *dst) {
+  // Either a field value or a variable.
+  if (bmv2_value.kind_case() != google::protobuf::Value::kStructValue) {
+    return absl::Status(absl::StatusCode::kInvalidArgument,
+                        "Left-hand of assignment is badly formatted!");
+  }
+
+  const google::protobuf::Struct &struct_value = bmv2_value.struct_value();
+  const std::string &type = struct_value.fields().at("type").string_value();
+  if (type == "field") {
+    const google::protobuf::ListValue &names =
+        struct_value.fields().at("value").list_value();
+
+    FieldValue *field_value = dst->mutable_field_value();
+    field_value->set_header_name(names.values(0).string_value());
+    field_value->set_field_name(names.values(1).string_value());
+  } else if (type == "runtime_data") {
+    int variable_index = struct_value.fields().at("value").number_value();
+
+    Variable *variable = dst->mutable_variable_value();
+    variable->set_name(variables[variable_index]);
+  } else {
+    return absl::Status(absl::StatusCode::kUnimplemented,
+                        "Unsupported expression in left-hand of assignment!");
+  }
+
+  return absl::OkStatus();
+}
+
+absl::Status TransformRValue(const google::protobuf::Value &bmv2_value,
+                             const std::vector<std::string> &variables,
+                             RValue *dst) {
+  // TODO(babman): Code duplication between this and TransformLValue.
+  //               This function will have more cases later when the
+  //               second todo is handled, but we still need to reduce
+  //               code duplication.
+  //               Difficulty: the type here is RValue instead of LValue.
+  //               Possible solution: create piece wise re-usable functions
+  //               that return FieldValue/Variable etc, cons: does not reduce
+  //               duplicates by much..
+  //               Will look at this later.
+  // TODO(babman): Support the remaining cases: literals and simple expressions.
+  if (bmv2_value.kind_case() != google::protobuf::Value::kStructValue) {
+    return absl::Status(absl::StatusCode::kInvalidArgument,
+                        "Left-hand of assignment is badly formatted!");
+  }
+
+  const google::protobuf::Struct &struct_value = bmv2_value.struct_value();
```

Comment:
Regarding your TODOs, looks like most everything below this is shared code?

Reply:
Yeah but there is a type issue